### PR TITLE
[TTT] Setup Firebase

### PR DESCRIPTION
**Pull request by Alonski**
This sets up Firebase in TTT using EmberFire and Ember Data addons

There were a few type errors related to `ember-data` and `emberfire`. I currently did a `patch-package` for them.

In this PR I also create a game in Firebase. Still not showing the games.
Also, I currently don't delete games on disconnect so games pile up in Firebase.
This will be handled in the next PR.

Issue for EmberFire: https://github.com/firebase/emberfire/issues/604
PR for Ember Data types: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43206

Fixes: #1143
Fixes: #1122

### CODE REVIEW

File: packages/xstate-wallet/stories/index.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
```
Comment (Alonski) [inline packages/xstate-wallet/stories/index.stories.tsx]: Should this file be deleted?

File: packages/xstate-wallet/src/messaging.ts [hunk 2]
```diff
@@ -177,6 +177,7 @@ export class MessagingService implements MessagingServiceInterface {
         break;
       case 'ChallengeChannel':
       case 'GetState':
+      case 'CloseAndWithdraw':
```
Comment (Alonski) [inline packages/xstate-wallet/src/messaging.ts]: Should this be reverted?

File: packages/wire-format/package.json [hunk 3]
```diff
@@ -10,6 +10,8 @@
   "devDependencies": {
     "@types/eslint": "6.1.7",
     "@types/eslint-plugin-prettier": "2.2.0",
+    "@types/jest": "25.1.0",
```
Comment (Alonski) [inline packages/wire-format/package.json]: Should this be reverted?

Comment (Alonski) [reply packages/wire-format/package.json]: Answering myself. I think not. In `tsconfig.json` it references these types

File: packages/tic-tac-toe/.env [hunk 4]
```diff
@@ -33,4 +33,4 @@ INFURA_API_KEY='A valid Infura API key is required to deploy to a test network'
 
 # These contract addresses get defined in the start/build scripts 
 # or the contract test setup with the actual addresses of the contracts
-TTT_CONTRACT_ADDRESS=0x0000000000000000000000000000000000000000
+TTT_CONTRACT_ADDRESS=0x430869383d611bBB1ce7Ca207024E7901bC26b40
```
Comment (snario) [inline packages/tic-tac-toe/.env]: This should probably be set in `.env.production` if this is a Ropsten address. Maybe also indicate the git commit has at which the contract was deployed from. Alternatively, verify the source code on etherscan.

File: packages/tic-tac-toe/app/controllers/games/index.ts [hunk 5]
```diff
@@ -1,16 +1,34 @@
...
-  @tracked buyInAmount = 0.1;
+  @service store!: DS.Store;
+  @service user!: UserService;
+
+  @tracked buyInAmount = 0.001;
   @tracked showCreateGameModal = false;
 
   @action
-  protected createNewGame(event: Event): void {
+  protected async createNewGame(event: Event): Promise<void> {
```
Comment (snario) [inline packages/tic-tac-toe/app/controllers/games/index.ts]: Why is this `async` now?

File: packages/tic-tac-toe/config/environment.js [hunk 6]
```diff
@@ -18,14 +18,20 @@ module.exports = function(environment) {
...
       // Here you can pass flags/options to your application instance
       // when it is created
+    },
+    firebase: {
+      apiKey: process.env.FIREBASE_API_KEY,
+      authDomain: `${process.env.FIREBASE_PROJECT}.firebaseapp.com`,
+      databaseURL: `https://${process.env.FIREBASE_PROJECT}.firebaseio.com`,
+      projectId: process.env.FIREBASE_PROJECT,
+      storageBucket: `https://${process.env.FIREBASE_PROJECT}.appspot.com`,
+      messagingSenderId: process.env.FIREBASE_MESSAGE_SENDER_ID
```
Comment (snario) [inline packages/tic-tac-toe/config/environment.js]: I don't think you need anything other than `databaseURL`

File: packages/tic-tac-toe/package.json [hunk 7]
```diff
@@ -113,13 +115,15 @@
...
     "ember-truth-helpers": "2.1.0",
+    "emberfire": "^3.0.0-rc.6",
     "enzyme": "3.11.0",
     "enzyme-adapter-react-16": "1.15.2",
     "eslint": "6.8.0",
     "eslint-plugin-ember": "7.7.2",
     "eslint-plugin-node": "11.0.0",
     "eslint-plugin-prettier": "3.1.2",
     "etherlime": "2.2.6",
+    "firebase": "7.11.0",
```
Comment (snario) [inline packages/tic-tac-toe/package.json]: Double check this is the same as the one in `rps`

File: packages/tic-tac-toe/.env (hunk 4 above)
Comment (Alonski) [reply packages/tic-tac-toe/.env]: I haven't deployed to Ropsten yet. Can you teach me how we do this?

Comment (snario) [reply packages/tic-tac-toe/.env]: What is this address then? You can probably use Truffle or Etherlime to deploy. I used `deploy.js` but modified the `GanacheDeployer` to one used for a real network when I did  it. I did not upload that code to version control.

Comment (Alonski) [reply packages/tic-tac-toe/.env]: This is the hard coded address from the local Gamache deploy. I had issues injecting the actual address during development so hard coding it for now so I don't get blocked.

File: packages/tic-tac-toe/app/controllers/games/index.ts (hunk 5 above)
Comment (Alonski) [reply packages/tic-tac-toe/app/controllers/games/index.ts]: I `await` the `newGame.save()` on line 31.

File: packages/tic-tac-toe/package.json (hunk 7 above)
Comment (Alonski) [reply packages/tic-tac-toe/package.json]: It is :)

File: packages/tic-tac-toe/config/environment.js (hunk 6 above)
Comment (Alonski) [reply packages/tic-tac-toe/config/environment.js]: In RPS we have all of these except an empty storageBucket.
```ts
const config = {
  apiKey: process.env.FIREBASE_API_KEY,
  authDomain: `${process.env.FIREBASE_PROJECT}.firebaseapp.com`,
  databaseURL: `https://${process.env.FIREBASE_PROJECT}.firebaseio.com`,
  projectId: process.env.FIREBASE_PROJECT,
  storageBucket: '',
  messagingSenderId: '913007764573',
};
```
https://github.com/statechannels/monorepo/blob/f98d3f2b2fb39905d533011a08ff1ef745f92b5b/packages/rps/src/gateways/firebase.ts#L7

Also note that I am using Firestore and not Realtime Database.

**Comment by Alonski** [comment]
@snario Just spent an hour trying to figure out how to setup firebase-server and stumbled upon this:
https://github.com/urish/firebase-server/issues/141

It doesn't work with Firestore.
RPS isn't using `firebase-server` either :)

**Comment by andrewgordstewart** [comment]
> RPS isn't using `firebase-server` either :)

Note that "using" `firebase-server` is as easy as setting the firebase URL to the URL of a local `firebase-server`. So, RPS can "use" `firebase-server` without depending on it -- if you are serving firebase at `http://localhost:9999` then RPS can use it by configuring the firebase URL to be `'http://localhost:9999'`.

**Comment by Alonski** [comment]
@andrewgordstewart
Thanks. As I said though I'm using Firestore which isn't supported by firebase server

File: packages/tic-tac-toe/.env (hunk 4 above)
Comment (snario) [reply packages/tic-tac-toe/.env]: @Alonski If you're following the same pattern as `rps` by the way this should be getting auto-injected by the deploy script.

Comment (snario) [reply packages/tic-tac-toe/.env]: If that's not working, I recommend putting this value into `.env.${NODE_ENV}.local` so it is not in version control.

File: packages/tic-tac-toe/config/environment.js (hunk 6 above)
Comment (snario) [reply packages/tic-tac-toe/config/environment.js]: Why are you using Firestore? `rps` uses the realtime database.

Comment (snario) [reply packages/tic-tac-toe/config/environment.js]: Also `rps` doesn't need any of those.

File: packages/tic-tac-toe/app/serializers/application.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import DS from 'ember-data';
+import FirestoreSerializer from 'emberfire/serializers/firestore';
+import {singularize} from 'ember-inflector';
+
+export default class ApplicationSerializer extends FirestoreSerializer {
+  normalizeCreateRecordResponse(
+    _store: DS.Store,
+    _primaryModelClass: DS.Model,
+    payload: {doc: {id: string; parent: {id: string}}; data: object},
+    id: string
```
Comment (spersico) [inline packages/tic-tac-toe/app/serializers/application.ts]: I'm not sure I get what this does. I'm assuming that it parses the response of `this.store.createRecord` from Firebase to the client, isn't it?

Comment (Alonski) [reply packages/tic-tac-toe/app/serializers/application.ts]: Added a comment about this.
There is an issue in latest EmberFire:
https://github.com/firebase/emberfire/pull/600

File: packages/tic-tac-toe/config/environment.js (hunk 6 above)
Comment (Alonski) [reply packages/tic-tac-toe/config/environment.js]: Done. Only need `FIREBASE_PROJECT` for:
`projectId` and `databaseURL`

**Pull request merged**